Adding new quantities (jetFlavour, HT)

fcolombo opened the pull request:
Hi Kappa users,

I created this pull request since the need of adding a couple of additional quantities to the Kappa dataformat showed up.
I started adding two new integer quantities for storing the jet flavour, relying on patJets methods, since apparently the way which was used so far is not optimal anymore.
@ffrensch could you please cross-check with Imperial that this is the quantity that they are actually using?I looked it up and these seemed the most reasonable to me, but better to be sure.
Since these methods are defined only for patJets, they are filled only in the KPatJetProducer, while in the KJetProducer they get default values. It seems to me that there is no other way around here, but please tell me if you have a better suggestion.

I will probably update this pull request ~soon to add another gen-level quantity (HT) which seems to be needed if one wants to reweight and stitch together some HT-samples which are available (see https://twiki.cern.ch/twiki/bin/viewauth/CMS/HiggsToTauTauWorking2015#MC_samples). I don't think that this information is available anywhere in Kappa, at the moment.

Normally these modifications should go in the dictchanges branch, since this is what they are. However, given that the master itself, right now, contains other dictchanges and we haven't tagged and done the next skim yet, I thought they could be added to the 2.1.0 release.

fcolombo committed:
Adding jetFlavour to KJet dataformat

FredStober commented:
If the HT value you spoke of is anything like pThat, it should be filled into KGenEventInfo::binValue

fcolombo commented:
Just to be sure that I understood. I agree that HT needs to be saved together with the other quantities in KGenEventInfo (if we need to have it at all). But are you suggesting this quantity is _already_ saved under the name "binValue", or that we should replace "binValue" by that quantity?
Is not actually clear to me what "binValue" should represent, and looking at this line it seems that Joram had the same impression:
https://github.com/KappaAnalysis/Kappa/blob/master/DataFormats/interface/KInfo.h#L114
At the moment, it simply gets a default value (-1) for us.

claria commented:
Hi,

i think binValue is just a general place to save a quantity which you need later
to distinguish between data samples to set cross section weights accordingly.
But do you really need this? Four the pthat/HT binned samples we just set the
cross section weight using the artus config based on the sample nickname, which
is much more easier and the relevant code in artus exists.

On 12/19/2015 03:01 PM, Fabio Colombo wrote:

> Just to be sure that I understood. I agree that HT needs to be saved together
> with the other quantities in KGenEventInfo (if we need to have it at all). But
> are you suggesting this quantity is /already/ saved under the name "binValue",
> or that we should replace "binValue" by that quantity?
> Is not actually clear to me what "binValue" should represent, and looking at
> this line it seems that Joram had the same impression:
> https://github.com/KappaAnalysis/Kappa/blob/master/DataFormats/interface/KInfo.h#L114
> At the moment, it simply gets a default value (-1) for us.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/KappaAnalysis/Kappa/pull/15#issuecomment-165986264.

fcolombo commented:
Hi, thanks for the clarification!
Yes, apart from code details, it is still not clear to me if we really need this quantity at all. You are right that for the HT-binned samples we would do it in Artus, exactly as you wrote. The question is more on how to use the inclusive sample.
My way of understanding it, is that we would still use the inclusive sample for the HT phase space which is **not** covered by the HT-binned ones. And for doing this, one needs to know the HT value of the event.
I need to check that this recipe is correct, though.

JoramBerger commented:
`binValue` is for samples which are produced in subsamples of different cross sections binned by some `binValue` (or `binningValue` in CMSSW) as Georg said before. Traditionally, I know of pthat as variable used here, e.g. for ZJet samples.

Jets are the most frequent objects per event usually. So please be careful with the size. Just saying. I am not sure whether that matters really, but `short` might be sufficient for jet flavours. But I would just use the data type used by CMSSW for it (the internal member might be smaller than the function return value). I think the jet flavour is a good addition.

In general I agree, that with the existing dict change, this is the time to include the quantities you need.

fcolombo committed:
Adding lheHt to KGenInfo dataformat

fcolombo commented:
The stitching recipe of the different samples is not so crystal-clear among the group, but it seems unavoidable to have the HT quantity available, to treat properly the events in the inclusive sample. The commit above adds this to the `KGenEventInfo`, at the expense of an additional double for each event.

Still to be done: check that what is given by the methods
`patJet::partonFlavour()`
`patJet::hadronFlavour()`
is what we want to have available in our next skims. @ajgilbert @ffrensch can you please confirm? Thanks a lot!

ffrensch commented:
Hi Fabio,

to my knowlegde |patJet::hadronFlavour() is the correct command (the 
other is older i believe).
I guess Andrew could confirm. Since he is in holidays i will ask back 
Rebecca.

Cheers,
Felix
|
On 21.12.2015 13:57, Fabio Colombo wrote:

> The stitching recipe of the different samples is not so crystal-clear 
> among the group, but it seems unavoidable to have the HT quantity 
> available, to treat properly the events in the inclusive sample. The 
> commit above adds this to the |KGenEventInfo|, at the expense of a 
> additional double for each event.
> 
> Still to be done: check that what is given by the methods
> |patJet::partonFlavour()|
> |patJet::hadronFlavour()|
> is what we want to have available in our next skims. @ajgilbert 
> https://github.com/ajgilbert @ffrensch https://github.com/ffrensch 
> can you please confirm? Thanks a lot!
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/KappaAnalysis/Kappa/pull/15#issuecomment-166298430.

ajgilbert commented:
Hi,

Yes hadronFlavour() is the correct one for determining the true flavour in the b-tag reweighting. See here for details:
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideBTagMCTools

The partonFlavour() is what was used in run 1 for the same purpose. No harm in keeping this still.

cheers,
Andrew

> On 21 Dec 2015, at 13:03, ffrensch notifications@github.com wrote:
> 
> Hi Fabio,
> 
> to my knowlegde |patJet::hadronFlavour() is the correct command (the 
> other is older i believe).
> I guess Andrew could confirm. Since he is in holidays i will ask back 
> Rebecca.
> 
> Cheers,
> Felix
> |
> On 21.12.2015 13:57, Fabio Colombo wrote:
> 
> > The stitching recipe of the different samples is not so crystal-clear 
> > among the group, but it seems unavoidable to have the HT quantity 
> > available, to treat properly the events in the inclusive sample. The 
> > commit above adds this to the |KGenEventInfo|, at the expense of a 
> > additional double for each event.
> > 
> > Still to be done: check that what is given by the methods
> > |patJet::partonFlavour()|
> > |patJet::hadronFlavour()|
> > is what we want to have available in our next skims. @ajgilbert 
> > https://github.com/ajgilbert @ffrensch https://github.com/ffrensch 
> > can you please confirm? Thanks a lot!
> > 
> > —
> > Reply to this email directly or view it on GitHub 
> > https://github.com/KappaAnalysis/Kappa/pull/15#issuecomment-166298430.
> 
> —
> Reply to this email directly or view it on GitHub.

ajgilbert commented:
And just to add:

Indeed the HT will be needed to combine the inclusive and binned HT samples. We don't want to throw away any events from the inclusive sample so we need to figure out the total number of events in each HT range that we have a cross section for (inclusive + exclusive) to determine the weight those events should receive. Technically speaking I think it will be easiest to apply this HT weighting as a per-event weight using pre-calculated event sums. The weight will be constructed such that sum of weights of all inclusive + exclusive events  = the unweighted number of inclusive events. After this you can treat all these as a single sample and just scale to the inclusive NNLO cross section as needed. 

cheers,
Andrew

> On 21 Dec 2015, at 13:14, Andrew Gilbert andrew.gilbert@cern.ch wrote:
> 
> Hi,
> 
> Yes hadronFlavour() is the correct one for determining the true flavour in the b-tag reweighting. See here for details:
> https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideBTagMCTools
> 
> The partonFlavour() is what was used in run 1 for the same purpose. No harm in keeping this still.
> 
> cheers,
> Andrew
> 
> > On 21 Dec 2015, at 13:03, ffrensch notifications@github.com wrote:
> > 
> > Hi Fabio,
> > 
> > to my knowlegde |patJet::hadronFlavour() is the correct command (the 
> > other is older i believe).
> > I guess Andrew could confirm. Since he is in holidays i will ask back 
> > Rebecca.
> > 
> > Cheers,
> > Felix
> > |
> > On 21.12.2015 13:57, Fabio Colombo wrote:
> > 
> > > The stitching recipe of the different samples is not so crystal-clear 
> > > among the group, but it seems unavoidable to have the HT quantity 
> > > available, to treat properly the events in the inclusive sample. The 
> > > commit above adds this to the |KGenEventInfo|, at the expense of a 
> > > additional double for each event.
> > > 
> > > Still to be done: check that what is given by the methods
> > > |patJet::partonFlavour()|
> > > |patJet::hadronFlavour()|
> > > is what we want to have available in our next skims. @ajgilbert 
> > > https://github.com/ajgilbert @ffrensch https://github.com/ffrensch 
> > > can you please confirm? Thanks a lot!
> > > 
> > > —
> > > Reply to this email directly or view it on GitHub 
> > > https://github.com/KappaAnalysis/Kappa/pull/15#issuecomment-166298430.
> > 
> > —
> > Reply to this email directly or view it on GitHub.

JoramBerger commented:
**flavour**: With the comments above, I now know that these are just the different flavour definitions. We can already calculate the physics and algorithmic definition in Artus. For the data format I would just use one new field `flavour` and fill it with the hadronFlavour. For any additional flavour definitions the tags should be used as  they are configurable (like we don't have different btags in the jet data format).

**HT**: But this sounds very much like the purpose of the `binValue` field. Why not use this then?

It makes sense to think about this twice. Considering Georg's situation, you can see how rare such dictchanges should be.

fcolombo commented:
**flavour**: yes, I'm aware of the possibility of calculating this at Artus level. The information that I had from @ffrensch however, was that this way of doing it was giving a poor agreement when sync'ing with Imperial. I haven't looked into this problem myself, so I can't tell much more about this, but I'm under the impression that something closer to the CMSSW definition is needed.
I'm in principle fine in adding only a `flavour` variable in the dataformats and to fill it with the most recent method.

**HT**: actually I'm not sure that `binValue` would help. Our use-case is that we have an "inclusive" sample containing events with different values of a certain variable (HT) and other "binned" samples which cover only a certain region in HT. We have to use all the events in all samples, weighting them according to the value of the variable HT (and therefore we need to know the per-event value assumed by that variable).
By the way, I don't know if I need to do something special, but what I see is that in my test skims `binValue` always gets a default value (-1), both when I run on an inclusive and on a binned sample

FredStober commented:
I would definitely fill the HT into the binValue variable, if that is the variable that is used to stitch the samples together - since thats the whole reason of being for the "binValue" variable.

The producer currently counts on the CMSSW GeneratorInterface filling the binningValue array correctly - since it seems it doesnt do the correct thing in your case, I would add some "binningMode" variable to the Producer and fill the binValue variable with what you need instead of the default -1.

Just adding HT and leaving binValue with the default value, would lead to the same situation in the long run, if you need to run over some MT binned or M binned sample at some point.

How exactly is the HT variable used while stiching together the phase space? Is it very different from the pThat, that is usually filled into that variable?

fcolombo committed:
Update in jet and event info dataformats
Only hadronFlavour method is used. Member binValue of the KInfo dataformat can be assigned to other event-based quantities, in a configurable way

fcolombo committed:
Small fix

fcolombo commented:
Back to this after the holiday break.
**flavour**: only a single new field (`flavour`) is added to the dataformat and contains the `hadronFlavour()` definition

**HT**: this quantity is now saved in the the already-existing variable named `binValue`, thus preserving the existing dataformat. There is some configurable parameter, in case in future a similar splitting will be done on other variables (mass, transverse mass, etc.)

fcolombo commented:
If there are no objections, I think that this PR could be merged, right? To summarize: the dataformat changes involves only the addition of `flavour` in the jets.
